### PR TITLE
Reduce invalid output of cmake usage heuristics

### PR DESCRIPTION
**Pull request by dg0yt**
For https://github.com/microsoft/vcpkg/issues/20190.

### CODE REVIEW

**Comment by dg0yt** [comment]
./vcpkg install grpc : 
~~~diff
 grpc provides CMake targets:
 
     # this is heuristically generated, and may not be correct
     find_package(gRPC CONFIG REQUIRED)
     # note: 7 additional targets are not displayed.
     target_link_libraries(main PRIVATE gRPC::gpr gRPC::grpc gRPC::grpc++ gRPC::grpc++_alts)
 
-    find_package(modules CONFIG REQUIRED)
-    target_link_libraries(main PRIVATE re2::re2 c-ares::cares)
-
~~~

**Comment by dg0yt** [comment]
xerces-c:
~~~diff
 xerces-c provides CMake targets:
 
     # this is heuristically generated, and may not be correct
     find_package(XercesC CONFIG REQUIRED)
-    target_link_libraries(main PRIVATE xerces_xerces-c XercesC::XercesC)
+    target_link_libraries(main PRIVATE XercesC::XercesC)

**Comment by dg0yt** [comment]
jsoncpp:
~~~diff
 jsoncpp provides CMake targets:
 
     # this is heuristically generated, and may not be correct
     find_package(jsoncpp CONFIG REQUIRED)
-    target_link_libraries(main PRIVATE jsoncpp_object jsoncpp_static JsonCpp::JsonCpp)
+    target_link_libraries(main PRIVATE JsonCpp::JsonCpp)

File: src/vcpkg-test/install.cpp [hunk 1]
```diff
@@ -123,18 +123,43 @@ cmake_policy(POP)
...
-
     res = get_cmake_add_library_names("add_library(foo) add_library(foo) add_library(foo)");
     CHECK(res == std::vector<std::string>{"foo", "foo", "foo"});
+
+    // In the following cases, the empty list indicates the need for an explicit usage file.
+
+    auto with_var = get_cmake_add_library_names("add_library(bar${foo}) add_library(${foo}) add_library(   \nbar)");
+    CHECK(with_var.empty());
+
+    auto with_comment = get_cmake_add_library_names("add_library( # rem \n    foo)  add_library( bar# rem3)");
```
Comment (Neumann-A) [inline src/vcpkg-test/install.cpp:140]:
```suggestion
    auto with_comment = get_cmake_add_library_names("add_library( # rem \n    foo)  add_library( bar# rem3\n)");
```

Comment (Neumann-A) [reply src/vcpkg-test/install.cpp:140]: or should the target name be bar# ?

Comment (dg0yt) [reply src/vcpkg-test/install.cpp:140]: The point is to not return any target name at all from this input. It is not the generated form. If there is a comment, it likely that the heuristics won't generate useful output.

Comment (Neumann-A) [reply src/vcpkg-test/install.cpp:140]: yeah but the input should still be valid cmake shouldn't it?

Comment (dg0yt) [reply src/vcpkg-test/install.cpp:140]: There is no benefit from perfection here. It is not a full parser.

**Comment by dg0yt** [comment]
Last not least (https://github.com/microsoft/vcpkg/issues/20190#issuecomment-1124558011):
~~~diff
-shapelib provides CMake targets:
-
-    # this is heuristically generated, and may not be correct
-    find_package(shp CONFIG REQUIRED)
-    target_link_libraries(main PRIVATE shp)
-

**Comment by dg0yt** [comment]
Header-only output due to lack of config file for exports:
~~~diff
-mdns provides CMake targets:
-
-    # this is heuristically generated, and may not be correct
-    find_package(cmake CONFIG REQUIRED)
-    target_link_libraries(main PRIVATE mdns::mdns)
+mdns is header-only and can be used from CMake via:
+
+    find_path(MDNS_INCLUDE_DIRS "mdns.h")
+    target_include_directories(main PRIVATE ${MDNS_INCLUDE_DIRS})
~~~

File: src/vcpkg/install.cpp [hunk 2]
```diff
@@ -740,122 +760,165 @@ namespace vcpkg
...
             }
-            else
+
+            if (has_targets_for_output)
             {
                 auto msg = msg::format(msgCMakeTargetsUsage, msg::package_name = bpgh.spec.name()).append_raw("\n\n");
                 msg.append_indent().append(msgCMakeTargetsUsageHeuristicMessage).append_raw('\n');
 
-                for (auto&& library_target_pair : library_targets)
+                for (const auto package_targets_pair : ret.cmake_targets_map)
```
Comment (BillyONeal) [inline src/vcpkg/install.cpp:874]:
```suggestion
                for (auto&& package_targets_pair : ret.cmake_targets_map)
```

I believe auto&& as formerly written is more correct here: there's no reason to introduce an extra by-value copy.

**Review by BillyONeal — APPROVED** [review]
Only extreme nitpicks here; this looks good to me :)

File: src/vcpkg/install.cpp [hunk 3]
```diff
@@ -740,122 +760,165 @@ namespace vcpkg
         auto files = fs.read_lines(installed.listfile_path(bpgh), ec);
         if (!ec)
         {
-            std::unordered_map<std::string, std::string> config_files;
+            // Mapping directories to unique config names and to library targets
+            struct ConfigPackages
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/install.cpp]: I'm suggesting leaving it in the function but out of the if.

**Comment by BillyONeal** [comment]
Thanks!

**Pull request merged**